File: backend/modules/petclinic/app/petclinic/api/pets/getPetDetail/GetPetDetailService.java
package petclinic.api.pets.getPetDetail;

import akka.actor.ActorRef;
import akka.pattern.AskSupport;
import akka.pattern.PatternsCS;
import petclinic.api.model.OwnerId;
import petclinic.api.model.PetType;
import petclinic.api.pettype.getAllPetTypes.GetAllPetTypeService;
import petclinic.model.OwnerMessages;

import javax.inject.Inject;
import javax.inject.Named;
import java.util.Collection;
import java.util.Optional;
import java.util.concurrent.CompletionStage;

import static petclinic.api.Constants.TIMEOUT;

public class GetPetDetailService implements AskSupport {

    @Inject
    @Named("owner")
    private ActorRef owner;

    @Inject
    private GetAllPetTypeService petTypeService;

    public CompletionStage<Optional<GetPetDetailApiResponse>> getPetDetail(OwnerId ownerId, String petId) {
        CompletionStage<Collection<PetType>> types = petTypeService.getAll();
        final CompletionStage<OwnerMessages.GetPetDetailsResponse> getPetDetailsResponse = PatternsCS.ask(owner, OwnerMessages.getPetDetail(ownerId.id(), petId), TIMEOUT)
                .thenApply(obj -> (OwnerMessages.GetPetDetailsResponse) obj);

        return types.thenCombine(getPetDetailsResponse, (petTypes, petDetails) -> {
            if (petDetails instanceof OwnerMessages.OwnerNotFound) {
                return Optional.empty();
            } else if (petDetails instanceof OwnerMessages.PetNotFound) {
                return Optional.empty();
            } else if (petDetails instanceof OwnerMessages.PetDetailsResponse) {
                OwnerMessages.PetDetailsResponse data = (OwnerMessages.PetDetailsResponse) petDetails;

                return Optional.of(GetPetDetailApiResponse.apply(
                        data.petId(),
                        data.petName(),
                        data.petBirthDay(),
                        petTypes.stream().filter(pt -> pt.getId().equals(data.petTypeId())).findFirst().get(), // if the type does not exist, we'll blow up.. should not happen anyway!
                        GetPetDetailApiResponseOwner.apply(data.ownerId(), data.ownerFirstName(), data.ownerLastName())
                ));
            } else {
                throw new IllegalStateException("should not happen");
            }
        });
    }


}

File: backend/modules/petclinic/app/petclinic/api/pets/getPetDetail/GetPetDetailService.scala
package petclinic.api.pets.getPetDetail

import akka.actor.ActorRef
import javax.inject.{Inject, Named}
import petclinic.api.Constants
import petclinic.api.model.OwnerId
import petclinic.api.pettype.getAllPetTypes.GetAllPetTypeService
import petclinic.model.owners.owner.Protocol._

import scala.concurrent.{ExecutionContext, Future}

class GetPetDetailService @Inject()(@Named(petclinic.api.Constants.OWNER_AGGREGATE_ACTOR_NAME) owner: ActorRef, getAllPetTypeService: GetAllPetTypeService)(implicit ec: ExecutionContext) {

  def getPetDetail(ownerIdIn: OwnerId, petIdIn: String): Future[Option[GetPetDetailApiResponse]] = {
    val typesFuture = getAllPetTypeService.getAll()
    val petDetailsFuture = akka.pattern.Patterns.ask(owner, GetPetDetailsQuery(ownerIdIn.id, petIdIn), Constants.TIMEOUT)
      .mapTo[GetPetDetailsResponse]

    typesFuture.zip(petDetailsFuture).map { case (petTypes, details) =>
      details match {
        case OwnerNotFoundGetPetDetailsResponse => None
        case PetNotFoundGetPetDetailsresponse => None
        case PetDetailsResponse(ownerId, ownerFirstName, ownerLastName, petId, petName, petBirthDay, petTypeId) =>
          Some(GetPetDetailApiResponse(
            petId,
            petName,
            petBirthDay,
            petTypes.find(x => x.getId.equals(petId)).get,
            GetPetDetailApiResponseOwner(ownerId, ownerFirstName, ownerLastName)
          ))
      }
    }

  }

}
